akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala

```diff
@@ -0,0 +1,467 @@
+  // discarded since they were in flight before the Resend request and will anyway be sent again.
+  private def resending(s: State[A]): Behavior[InternalCommand] = {
+    Behaviors.receiveMessage {
+      case seqMsg: SequencedMessage[A] =>
+        val pid = seqMsg.producerId
+        val seqNr = seqMsg.seqNr
+
+        if (s.isProducerChanged(seqMsg)) {
+          receiveChangedProducer(s, seqMsg)
+        } else if (s.registering.isDefined) {
+          context.log.infoN(
+            "from producer [{}], discarding message because registering to new ProducerController, seqNr [{}]",
+            pid,
+            seqNr)
+          Behaviors.same
+        } else if (s.isNextExpected(seqMsg)) {
+          context.log.infoN("from producer [{}], received missing [{}]", pid, seqNr)
+          deliver(s.copy(receivedSeqNr = seqNr, registering = s.updatedRegistering(seqMsg)), seqMsg)
+        } else {
+          context.log.infoN("from producer [{}], ignoring [{}], waiting for [{}]", pid, seqNr, s.receivedSeqNr + 1)
+          if (seqMsg.first)
+            retryRequest(s)
+          Behaviors.same // ignore until we receive the expected
+        }
+
+      case Retry =>
+        receiveRetry(s, () => {
+          // in case the Resend message was lost
+          context.log.info("retry Resend [{}]", s.receivedSeqNr + 1)
+          s.producer ! Resend(fromSeqNr = s.receivedSeqNr + 1)
+          Behaviors.same
+        })
+
+      case Confirmed(seqNr) =>
+        receiveUnexpectedConfirmed(seqNr)
+
+      case start: Start[A] =>
+        receiveStart(s, start, newState => resending(newState))
+
+      case ConsumerTerminated(c) =>
+        receiveConsumerTerminated(c)
+
+      case reg: RegisterToProducerController[A] =>
+        receiveRegisterToProducerController(s, reg, newState => active(newState))
+
+      case _: UnsealedInternalCommand =>
+        Behaviors.unhandled
+    }
+  }
+
+  private def deliver(s: State[A], seqMsg: SequencedMessage[A]): Behavior[InternalCommand] = {
+    context.log.info("from producer [{}], deliver [{}] to consumer", seqMsg.producerId, seqMsg.seqNr)
+    s.consumer ! Delivery(seqMsg.producerId, seqMsg.seqNr, seqMsg.msg, context.self)
+    waitingForConfirmation(s, seqMsg)
+  }
+
+  // The message has been delivered to the consumer and it is now waiting for Confirmed from
+  // the consumer. New SequencedMessage from the ProducerController will be stashed.
+  private def waitingForConfirmation(s: State[A], seqMsg: SequencedMessage[A]): Behavior[InternalCommand] = {
+    Behaviors.receiveMessage {
+      case Confirmed(seqNr) =>
+        val expectedSeqNr = s.receivedSeqNr
+        if (seqNr > expectedSeqNr) {
+          throw new IllegalStateException(
+            s"Expected confirmation of seqNr [$expectedSeqNr], but received higher [$seqNr]")
+        } else if (seqNr != expectedSeqNr) {
+          context.log.info(
```

Comment:
Maybe I was thinking that something could be in flight from a restart, but then it should first be reset via a new `Start`. Changed so that it's only accepting the expected.